config/v1/types_ingress.go

```diff
@@ -41,13 +43,122 @@ type IngressSpec struct {
 	// certificate.
 	// +optional
 	AppsDomain string `json:"appsDomain,omitempty"`
+
+	// componentRoutes is an optional list of routes configured by the cluster-admin to
+	// override the default hostname and serving certificates. The namespace and name of
+	// each componentRoute should match an existing componentRoute defined in the status.
+	//
+	// To determine the set of configurable Routes, look at namespace and name of entries in the
+	// .status.componentRoutes, where participating operators place current route status.
+	// +optional
+	ComponentRoutes []ComponentRouteSpec `json:"componentRoutes,omitempty"`
 }
 
+// consumingUser is an alias for string which we add validation to. Currently only service accounts are supported.
+// +kubebuilder:validation:Pattern="^system:serviceaccount:[a-z0-9]([-a-z0-9]*[a-z0-9])?:[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=512
+type consumingUser string
+
+// hostname is an alias for hostname string validation.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=253
+// +kubebuilder:validation:Format=hostname
+type hostname string
```

Comment:
Elsewhere, we use a string type with kubebuilder validation tags on the field.  Is specifying min and max lengths necessary when format is already specified as hostname?  Is defining a new type useful here?

Reply:
You cannot set the min/max length for strings that appear in an array, the existing workaround for string arrays that require min/max lengths is to create an alias.

Reply:
I would have expected `+kubebuilder:validation:Format=hostname` to suffice since an empty string or a string longer than 253 characters is not a valid host name.  But maybe the same issue (that the validation needs to apply to the elements and not the entire slice) applies to `+kubebuilder:validation:Format=hostname`?

Reply:
After looking into this, the format=hostname forces a limited length of 1 character and a maximum length of 256 characters. Unfortunately, applying the pattern to an array does not prevent invalid hostnames from being declared in the list. I have removed the min and max length specifications though.